sweepbatcher/presigned.go

```diff
@@ -0,0 +1,457 @@
+	// all the invariants.
+	err = CheckSignedTx(tx, signedTx, batchAmt, minRelayFee)
+	if err != nil {
+		return 0, fmt.Errorf("signed tx doesn't correspond the "+
+			"unsigned tx: %w", err), false
+	}
+	tx = signedTx
+	txHash := tx.TxHash()
+
+	// Make sure tx weight matches the expected value.
+	realWeight := lntypes.WeightUnit(
+		blockchain.GetTransactionWeight(btcutil.NewTx(tx)),
+	)
+	if realWeight != weight {
+		b.Warnf("actual weight of tx %v is %v, estimated as %d",
+			txHash, realWeight, weight)
+	}
+
+	// Find actual fee rate of the signed transaction. It may differ from
+	// the desired fee rate, because SignTx may return a presigned tx.
+	output := btcutil.Amount(tx.TxOut[0].Value)
+	fee = batchAmt - output
+	signedFeeRate := chainfee.NewSatPerKWeight(fee, realWeight)
+
+	numSweeps := len(tx.TxIn)
+	b.Infof("attempting to publish custom signed tx=%v, desiredFeerate=%v,"+
+		" signedFeeRate=%v, weight=%v, fee=%v, sweeps=%d, destAddr=%s",
+		txHash, feeRate, signedFeeRate, weight, fee, numSweeps, address)
+	b.debugLogTx("serialized batch", tx)
+
+	// Publish the transaction.
+	err = b.wallet.PublishTransaction(ctx, tx, b.cfg.txLabeler(b.id))
+	if err != nil {
+		return 0, fmt.Errorf("publishing tx failed: %w", err), true
+	}
+
+	// Store the batch transaction's txid and pkScript, for monitoring
+	// purposes.
+	b.batchTxid = &txHash
+	b.batchPkScript = tx.TxOut[0].PkScript
+
+	return fee, nil, true
+}
+
+// getSweepsDestAddr returns the destination address used by a group of sweeps.
+// The method must be used in presigned mode only.
+func (b *batch) getSweepsDestAddr(ctx context.Context,
+	sweeps []sweep) (btcutil.Address, error) {
+
+	if b.cfg.presignedHelper == nil {
+		return nil, fmt.Errorf("getSweepsDestAddr used without " +
+			"presigned mode")
+	}
+
+	inputs := make([]wire.OutPoint, len(sweeps))
+	for i, s := range sweeps {
+		if !s.presigned {
+			return nil, fmt.Errorf("getSweepsDestAddr used on a " +
+				"non-presigned input")
+		}
+
+		inputs[i] = s.outpoint
+	}
+
+	// Load pkScript from the presigned helper.
+	pkScriptBytes, err := b.cfg.presignedHelper.DestPkScript(ctx, inputs)
+	if err != nil {
+		return nil, fmt.Errorf("presignedHelper.DestPkScript failed "+
+			"for inputs %v: %w", inputs, err)
+	}
+
+	// Convert pkScript to btcutil.Address.
+	pkScript, err := txscript.ParsePkScript(pkScriptBytes)
+	if err != nil {
+		return nil, fmt.Errorf("txscript.ParsePkScript failed for "+
+			"pkScript %x returned for inputs %v: %w", pkScriptBytes,
+			inputs, err)
+	}
+
+	address, err := pkScript.Address(b.cfg.chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("pkScript.Address failed for "+
+			"pkScript %x returned for inputs %v: %w", pkScriptBytes,
+			inputs, err)
+	}
+
+	return address, nil
+}
+
+// CheckSignedTx makes sure that signedTx matches the unsignedTx. It checks
+// according to criteria specified in the description of PresignedHelper.SignTx.
```

Comment:
looks like the input sequence numbers aren't checked as described in the godoc of `SignTx`.

Reply:
Sequence numbers are compared using `unsignedMap`. Also there is a test case to verify this.